pkg/ipcache/metadata.go

```diff
@@ -0,0 +1,378 @@
+		logfields.CIDR:   cidr,
+		logfields.Labels: lbls, // omitting allLbls as CIDR labels would make this massive
+	}).Debug(
+		"Injecting CIDR labels for prefix",
+	)
+
+	return allocate(cidr, allLbls)
+}
+
+// FilterMetadataByLabels returns all the prefixes inside the IdentityMetadata
+// map which contain the given labels. Note that `filter` is a subset match,
+// not a full match.
+func FilterMetadataByLabels(filter labels.Labels) []string {
+	idMDMU.RLock()
+	defer idMDMU.RUnlock()
+	matching := make([]string, 0)
+	for prefix, lbls := range IdentityMetadata {
+		if bytes.Contains(lbls.SortedList(), filter.SortedList()) {
+			matching = append(matching, prefix)
+		}
+	}
+	return matching
+}
+
+// RemoveAllPrefixesWithLabels wraps RemoveLabels to provide a convenient
+// method for the caller to remove all given prefixes at once. This function
+// will trigger policy update and recalculation if necessary on behalf of the
+// caller if any changes to the kube-apiserver were detected.
+//
+// Identities allocated by InjectLabels() may be released by RemoveLabels().
+//
+// A prefix will only be removed from the IDMD if the set of labels becomes
+// empty.
+func RemoveAllPrefixesWithLabels(
+	m map[string]labels.Labels,
+	src source.Source,
+	updater identityUpdater,
+	triggerer policyTriggerer,
+) {
+	var (
+		hasKubeAPIServer bool
+
+		idsToPropagate = make(map[identity.NumericIdentity]labels.LabelArray)
+	)
+	for prefix, lbls := range m {
+		id, exists := IPIdentityCache.LookupByIP(prefix)
+		if id.ID <= identity.IdentityUnknown || !exists {
+			continue
+		}
+		if lbls.Has(labels.LabelKubeAPIServer[labels.IDNameKubeAPIServer]) {
+			hasKubeAPIServer = true
+		}
+		// Insert to propagate the updated set of labels after removal.
+		idsToPropagate[id.ID] = RemoveLabels(prefix, lbls, src).LabelArray()
+	}
+	if hasKubeAPIServer {
+		if updater == nil || triggerer == nil {
+			log.Warn("Unable to trigger policy updates after removing from ipcache because policy subsystem is not yet initialized.")
+			return
+		}
+
+		var wg sync.WaitGroup
+		updater.UpdateIdentities(nil, idsToPropagate, &wg)
+		wg.Wait()
+
+		triggerer.TriggerPolicyUpdates(false, "kube-apiserver identity updated by removal")
+	}
+}
+
+// RemoveLabels removes the given labels association with the given prefix. The
+// leftover labels are returned, if any.
+//
+// Identities are deallocated and their subequent entry in the IPCache is
+// removed if the prefix is no longer associated with any labels.
+//
+// It is the responsibility of the caller to trigger policy recalculation after
+// calling this function.
+func RemoveLabels(prefix string, lbls labels.Labels, src source.Source) labels.Labels {
+	idMDMU.Lock()
+	defer idMDMU.Unlock()
+
+	l, ok := IdentityMetadata[prefix]
+	if !ok {
+		return nil
+	}
+
+	l.Remove(lbls)
+	if !l.Equals(lbls) { // Labels left over, do not deallocate
+		n := labels.NewLabelsFromModel(nil)
+		n.MergeLabels(l)
+		return n // copy of leftover
```

Comment:
The identity handling here is a bit fishy - if the IdentityMetadata labels set has changed and identities are generally tied to the lifetime of the entry in the IdentityMetadata map, then this case should deallocate the identity for the previous set of labels and allocate the identity for the new set of labels.